storage/postgres/log_storage.go

```diff
@@ -0,0 +1,999 @@
+		}
+	}
+	totalDuration := time.Since(start)
+	readDuration := totalDuration - insertDuration
+	observe(queueReadLatency, readDuration, label)
+	observe(queueLatency, totalDuration, label)
+
+	return existingLeaves, nil
+}
+
+func (t *logTreeTX) AddSequencedLeaves(ctx context.Context, leaves []*trillian.LogLeaf, timestamp time.Time) ([]*trillian.QueuedLogLeaf, error) {
+	res := make([]*trillian.QueuedLogLeaf, len(leaves))
+	ok := status.New(codes.OK, "OK").Proto()
+
+	// Leaves in this transaction are inserted in two tables. For each leaf, if
+	// one of the two inserts fails, we remove the side effect by rolling back to
+	// a savepoint installed before the first insert of the two.
+	const savepoint = "SAVEPOINT AddSequencedLeaves"
+	if _, err := t.tx.ExecContext(ctx, savepoint); err != nil {
+		glog.Errorf("Error adding savepoint: %s", err)
+		return nil, err
+	}
+	// TODO(pavelkalinnikov): Consider performance implication of executing this
+	// extra SAVEPOINT, especially for 1-entry batches. Optimize if necessary.
+
+	// Note: LeafData inserts are presumably protected from deadlocks due to
+	// sorting, but the order of the corresponding SequencedLeafData inserts
+	// becomes indeterministic. However, in a typical case when leaves are
+	// supplied in contiguous non-intersecting batches, the chance of having
+	// circular dependencies between transactions is significantly lower.
+	ordLeaves := sortLeavesForInsert(leaves)
+	for _, ol := range ordLeaves {
+		i, leaf := ol.idx, ol.leaf
+
+		// This should fail on insert, but catch it early.
+		if got, want := len(leaf.LeafIdentityHash), t.hashSizeBytes; got != want {
+			return nil, status.Errorf(codes.FailedPrecondition, "leaves[%d] has incorrect hash size %d, want %d", i, got, want)
+		}
+
+		if _, err := t.tx.ExecContext(ctx, savepoint); err != nil {
+			glog.Errorf("Error updating savepoint: %s", err)
+			return nil, err
+		}
+
+		res[i] = &trillian.QueuedLogLeaf{Status: ok}
+
+		// TODO(pavelkalinnikov): Measure latencies.
+		_, err := t.tx.ExecContext(ctx, insertLeafDataSQL,
+			t.treeID, leaf.LeafIdentityHash, leaf.LeafValue, leaf.ExtraData, timestamp.UnixNano())
+		// TODO(pavelkalinnikov): Detach PREORDERED_LOG integration latency metric.
+
+		// TODO(pavelkalinnikov): Support opting out from duplicates detection.
+		//if isDuplicateErr(err) {
+		//		res[i].Status = status.New(codes.FailedPrecondition, "conflicting LeafIdentityHash").Proto()
+		// Note: No rolling back to savepoint because there is no side effect.
+		//		continue
+		//} else
+		if err != nil {
+			glog.Errorf("Error inserting leaves[%d] into LeafData: %s", i, err)
+			return nil, err
+		}
+
+		dupCheckRow, err := t.tx.QueryContext(ctx, insertSequencedLeafSQL,
+			t.treeID, leaf.LeafIndex, leaf.LeafIdentityHash, leaf.MerkleLeafHash, 0)
+		// TODO(pavelkalinnikov): Update IntegrateTimestamp on integrating the leaf.
+		resultData := true
+		for dupCheckRow.Next() {
+			dupCheckRow.Scan(&resultData)
```

Comment:
strictly, you should probably be checking the returned `error` here...

Reply:
Sure thing